tests/data/format/capitalization_first_letter/function_docstrings.py

```diff
@@ -0,0 +1,19 @@
+def func():
+    """
+    a docstring"""
+
+    def inner_func():
```

Comment:
Could you add some tests for:
```python
def func(named_parameter: str) -> None:
    """named_parameter should have a nice name
    
    Other information here
    """
    print(named_parameter)
```

Whenever the sentence starts with one of the parameter names I don't think we should enforce capitalisation.

Reply:
Did not thougth of that, nice catch. Shouldn't a parameter name be inside ``'`` ? I think we should detect it anyway but maybe the result could be made from another Formatter :
```
def func(named_parameter: str) -> None:
    """'named_parameter' should have a nice name
    
    Other information here
    """
    print(named_parameter)
```  ?

Reply:
If you look for `Do not use the Emacs convention of mentioning the arguments of functions or methods in upper case in running text.` in https://www.python.org/dev/peps/pep-0257/ you can find the (I think) only mention of argument names.
I'm on the fence about the `'`. If you want to enforce that a multi-line docstring starts with a single sentence and also enforce a short max line length adding `'` can eat into your precious character count..

Note that PEP 8 says that for docstrings the max length is 72... "For flowing long blocks of text with fewer structural restrictions (docstrings or comments), the line length should be limited to 72 characters." at https://www.python.org/dev/peps/pep-0008/


I'd say we can make enforcing `'` an optional check and for now only focus on non-capitalisation of parameter names.

Reply:
We might have to give the previous token(s?) to the Formatters. We need more than the docstring token if we need to get the name of the function argument. We can add optional argument to ``treat_token`` for all formatters, or initialize the Formatters with optional value for all formatters, or do a specific treatment for ``ContextAwareFormatter``... any ideas welcome :)

Reply:
I like `ContextAwareFormatter` as it clearly differentiates different types of formatters. However, I worry how it would like in `_format_file`. Do we need `formatting.FORMATTERS` and `formatting.CONTEXT_AWARE_FORMATTERS` then?

Adding an optional argument to `treat_token` might make more sense, but we need to think about what to pass then:
only one token or a list of tokens? And if we pass a list of tokens, how are we going to now which tokens to include in that list?

Reply:
It might be easier not to use `FORMATTERS` but call each formatter in `_format_file` individually. That gives us some more freedom in choosing what to pass to each formatter.

Reply:
> Do we need formatting.FORMATTERS and formatting.CONTEXT_AWARE_FORMATTERS then?

I was thinking of ``if is_instance(formatter, ContextAwareFormatter)`` but this hurts the liskov substitution princple. It's still less hacky than detecting snake_case with regex like I did in the latest commit though :troll:. Maybe two lists is better.

Reply:
Alternatively maybe instantiating the tokens in the Formatters is not that costly performance wise and we could store context in formatters all the time ?

Reply:
One of the problems I foresee is if we want to provide different `context` to different formatters.
What if one needs the previous tokens until the last `class/def` call while the other wants to see if there is a new line in between the docstring and the `__init__` declaration (the latter is a PEP recommendation)?

Reply:
> Alternatively maybe instantiating the tokens in the Formatters is not that costly performance wise and we could store context in formatters all the time ?

Based on my experience with `pylint` I know that iterating over `tokens` can be very costly. Remember that PR about a thousand empty tuples?

I think it makes more sense to ditch the `FORMATTERS` list and just call each formatter independently. If we want to keep the call small we could even combine formatters in a parent class and call that class.

Reply:
Indeed if we're going to create a special case for each specific token we need to give it's defeating the purpose of the design in the first place. If we give all the tokens all the time it would solves the issue. Do you want me to do some performance check ?

Reply:
Not sure if it defeats the purpose of the design. I think the small size of the `BeginningQuotesFormatter` shows how effective it already is.

We can see if passing all tokens works. It's probably easier to pass all tokens now and narrow it down later if we want to then create a design based on subsets of tokens that gradually expands.

Reply:
> we could even combine formatters in a parent class and call that class.

I like the idea of combining all the formatters that needs token in the same class. So we could have a list of `SIMPLE_FORMATTERS` and a class with all the context aware formatters.

Reply:
Btw, just getting ahead of myself: we'll likely also need to pass `self.config` at some point to allow optional formatters to see if they should run.. 😅

Reply:
I have a context aware formatter ready but maybe this hack with regex is enough, let me know :)

Reply:
This could be really stupid (and I'm not on regex101 right now), but what about:
```python
def func(param):
    """param is a nice parameter"""
```

This regex doesn't catch that right?